mypy/nodes.py

```diff
@@ -2234,21 +2234,55 @@ def __getattribute__(self, attr: str) -> None:
 
 
 class SymbolTableNode:
+    """Description of a name binding in a symbol table.
+
+    These are only used as values in module (global), function (local)
+    and class symbol tables (see SymbolTable). The name that is bound is
+    the key in SymbolTable.
+
+    Symbol tables don't contain direct references to AST nodes primarily
+    because there can be multiple symbol table references to a single
+    AST node (due to imports and aliases), and different references can
+    behave differently. This class describes the unique properties of
+    each reference.
+
+    The most fundamental attributes are 'kind' and 'node'.  The 'node'
+    attribute defines the AST node that the name refers to.
+
+    For many bindings, including those targeting variables, functions
+    and classes, the kind is one of LDEF, GDEF or MDEF, depending on the
+    scope of the definition. These three kinds can usually be used
+    interchangeably and the difference between local, global and class
+    scopes is mostly descriptive, with no semantic significance.
+    However, some tools that consume mypy ASTs may care about these so
+    they should be correct.
+
+    A few definitions get special kinds, including type variables (TVAR),
+    imported modules and module aliases (MODULE_REF), and type aliases
+    (TYPE_ALIAS).
+
+    Type aliases are very special and have additional attributes that
+    are only used for them ('type_override', 'alias_tvars' at least).
+    """
+    # TODO: This is a mess. Refactor!
+    # TODO: Describe how type aliases work.
```

Comment:
Type aliases are quite complicated (due to several corner cases). I tried to make some simplifications recently. It seems to me that we can introduce a dedicated ``SybmolNode`` (like ``Var`` or ``TypeInfo``) and deprecate ``type_override`` and friends and just point ``.node`` to this new symbol node. This is major refactoring, but I think it may help.

Reply:
Yeah, that's what I was thinking about. I'd also like to deprecate the special handling of module references and other special node kinds so that we could remove the `kind` attribute and replace it with a scope attribute, with only 3 possible values, corresponding to `LDEF`, `GDEF` and `MDEF`. Not sure how feasible this would be. It would be quite a big refactoring in any case.

Reply:
Maybe it makes sense to open an issue to track this?

Reply:
Created https://github.com/python/mypy/issues/4082.